### PR TITLE
Fix Bug 1140186 - Update meta tags on Firefox pages

### PR DESCRIPTION
**Pull request by kyoshino**

### CODE REVIEW

File: bedrock/firefox/templates/firefox/developer.html [hunk 1]
```diff
@@ -6,7 +6,16 @@
 
 {% block page_title_prefix %}{% endblock %}
 {% block page_title %}{{_('Firefox Developer Edition')}}{% endblock %}
-{% block page_desc %}{{_('Download Mozilla Firefox, a free Web browser. Firefox is created by a global non-profit dedicated to putting individuals in control online. Get Firefox today!')}}{% endblock %}
+{% block page_image %}{{ static('img/firefox/developer/page-image.png') }}{% endblock %}
+
+{% block page_desc %}
+  {% if l10n_has_tag('meta_update_201403') %}
```
Comment (flodolo) [inline bedrock/firefox/templates/firefox/developer.html]: You really need to convince yourself that we're in 2015 ;-)

Comment (kyoshino) [inline bedrock/firefox/templates/firefox/developer.html]: Ahhhhh. I just don't want to get old  :stuck_out_tongue_closed_eyes:

File: bedrock/firefox/templates/firefox/privacy_tour/index.html [hunk 2]
```diff
@@ -11,6 +11,20 @@
 
 {% block extra_meta %}<meta name="robots" content="noindex">{% endblock %}
 
+{% block page_title %}
```
Comment (alexgibson) [inline bedrock/firefox/templates/firefox/privacy_tour/index.html]: I'm a bit confused as to why we're updating strings on this page? The privacy tour was shown as a whatsnew page for users updating to 33.1, which shipped quite a while ago. I doubt there's much traffic to this page (if any at all). It's also a page only viewable by FF users and is not indexed by search engines, so I'm not sure what benefit there is here.

Comment (alexgibson) [inline bedrock/firefox/templates/firefox/privacy_tour/index.html]: Come to think of it, would the `firefox/independent` page be a better fit for this, since that was the publically viewable campaign page for this release?

File: bedrock/firefox/templates/firefox/dev-firstrun.html [hunk 3]
```diff
@@ -15,6 +15,8 @@
 {% block page_title_prefix %}{% endblock %}
 {% block page_title %}{{_('Welcome to Firefox Developer Edition')}}{% endblock %}
 {% block page_title_suffix %}{% endblock %}
+{% block page_image %}{{ static('img/firefox/developer/page-image.png') }}{% endblock %}
```
Comment (alexgibson) [inline bedrock/firefox/templates/firefox/dev-firstrun.html]: There's no harm in adding this image here, but `/firstrun` pages should not form part of a social campaign, since they are treated as in-product pages only accessible to Firefox. Probably good to make sure folks understand this when asking for changes.

File: bedrock/firefox/templates/firefox/all.html [hunk 4]
```diff
@@ -13,7 +13,15 @@
   {% elif channel == 'esr' %}
     {{ _('Firefox ESR is intended for groups who deploy and maintain the desktop environment in large organizations. <a href="%s">Learn more.</a>')|format(url('firefox.organizations.organizations')) }}
   {% else %}
-    {{ _('Firefox is available in more than 80 languages thanks to the contributions of Mozilla community members from around the world.') }}
+    {{ _('Firefox is made in large part by volunteers around the world. That’s why we offer it in 80 languages (and why there’s heart and soul in every piece of code).') }}
```
Comment (alexgibson) [inline bedrock/firefox/templates/firefox/all.html]: This is a new string not in an l10n tag, but I'm assuming this is ok and you are handling the `ja` locale?

The current one is still untranslated on prod since the last change: https://www.mozilla.org/ja/firefox/all/

Comment (flodolo) [inline bedrock/firefox/templates/firefox/all.html]: Japanese is the only localization for this page, I assume @kyoshino is already on this one?

File: bedrock/firefox/templates/firefox/privacy_tour/index.html (hunk 2 above)
Comment (kyoshino) [inline bedrock/firefox/templates/firefox/privacy_tour/index.html]: Yes, the whatnew pages are not intended to be shared. I'll ask in the bug.

File: bedrock/firefox/templates/firefox/all.html (hunk 4 above)
Comment (kyoshino) [inline bedrock/firefox/templates/firefox/all.html]: Yes, I can update the Japanese strings myself.

**Comment by alexgibson** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1140186#c8

"Please don't update the title on /firefox/new/ as that one has already been optimizations for SEO. All other pages, I am fine with. /firefox/new/ is critical to optimize to seo and not brand terms."

File: bedrock/firefox/templates/firefox/new.html [hunk 5]
```diff
@@ -6,7 +6,22 @@
 
 {% block page_title_prefix %}{{_('Download Firefox')}} — {% endblock %}
 {% block page_title %}{{_('Free Web Browser')}}{% endblock %}
-{% block page_desc %}{{_('Download Mozilla Firefox, a free Web browser. Firefox is created by a global non-profit dedicated to putting individuals in control online. Get Firefox today!')}}{% endblock %}
+
+{#- This will appear as <meta property="og:title"> which can be used for social share -#}
```
Comment (alexgibson) [inline bedrock/firefox/templates/firefox/new.html]: https://bugzilla.mozilla.org/show_bug.cgi?id=1140186#c8

"Please don't update the title on /firefox/new/ as that one has already been optimizations for SEO. All other pages, I am fine with. /firefox/new/ is critical to optimize to seo and not brand terms."

File: bedrock/firefox/templates/firefox/privacy_tour/index.html (hunk 2 above)
Comment (alexgibson) [inline bedrock/firefox/templates/firefox/privacy_tour/index.html]: I would recommend removing this from the PR, as it's just creating superfluous changes.

File: bedrock/firefox/templates/firefox/new.html (hunk 5 above)
Comment (kyoshino) [inline bedrock/firefox/templates/firefox/new.html]: The `page_og_title` block is only for `<meta property="og:title">`. The `<title>` will remain as is.

**Comment by kyoshino** [comment]
Discarded the changes to the firstrun and whatsnew pages. Nothing harmful but useless.

File: bedrock/firefox/templates/firefox/new.html (hunk 5 above)
Comment (alexgibson) [inline bedrock/firefox/templates/firefox/new.html]: The page description is also changing here though? Will this likely also impact SEO on this page?

Comment (kyoshino) [inline bedrock/firefox/templates/firefox/new.html]: Ahh. That's why I myself have introduced the `page_og_desc` block as well...

File: bedrock/firefox/templates/firefox/whatsnew-base-resp.html [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+{#- The firstrun and whatsnew pages should not be indexed by search engines -#}
+{% block extra_meta %}<meta name="robots" content="noindex">{% endblock %}
+
+{#- Override <meta property="og:url"> for social share -#}
+{% block page_og_url %}{{ url('firefox') }}{% endblock %}
+
+{#- This will appear as <meta property="og:description"> which can be used for social share -#}
+{% block page_og_desc %}
+  {% if l10n_has_tag('meta_update_201503') %}
+    {{_('Firefox is non-profit, non-corporate, non-compromised. Choosing Firefox isn’t just choosing a browser. It’s a vote for personal freedom.')}}
```
Comment (flodolo) [inline bedrock/firefox/templates/firefox/whatsnew-base-resp.html]: This doesn't make sense. Why are we adding this string to a base template? Base template should not rely on l10n tags either.

Same for the next file. If it's something we need, we'll have to add it to a shared file like main.lang

Comment (alexgibson) [inline bedrock/firefox/templates/firefox/whatsnew-base-resp.html]: I'm confused why we're now adding these to all whatsnew pages? As previously mentioned, these should be treated as internal FF only pages and are not really meant for sharing. Much of the code for many of these pages is only targeted for specific versions of Firefox.

Comment (kyoshino) [inline bedrock/firefox/templates/firefox/whatsnew-base-resp.html]: Looks like I have force pushed my WIP, but... because there are many whatsnew and firstrun templates and I didn't want to add the string to each template. That string can go to `main.lang` I guess.

Comment (alexgibson) [inline bedrock/firefox/templates/firefox/whatsnew-base-resp.html]: Kohei, are these changes really necessary in these pages?

Comment (kyoshino) [inline bedrock/firefox/templates/firefox/whatsnew-base-resp.html]: @alexgibson according to discussion in the bug, those in-product URLs may also be shared by visitors even though those are dead-end pages, so I have proposed a solution of overriding `og:url` to make the pages shareable.

Comment (flodolo) [inline bedrock/firefox/templates/firefox/whatsnew-base-resp.html]: To be completely honest, I don't like the approach. We can add the string to main.lang, and then we're stuck next time we decide to update this string because we want to focus communication on something else.

I'd prefer what's new page/first run to be threated individually, and add string as necessary. I'd also like to see numbers on how many people actually share the whatsnew page (do we have them?), because it really seems an edge case.

Comment (alexgibson) [inline bedrock/firefox/templates/firefox/whatsnew-base-resp.html]: Regardless as to whether some people may share them, I don't think it is something we should necessarily encourage :/

Comment (alexgibson) [inline bedrock/firefox/templates/firefox/whatsnew-base-resp.html]: I've commented in the bug for discussion, thanks @kyoshino

Comment (kyoshino) [inline bedrock/firefox/templates/firefox/whatsnew-base-resp.html]: Um, yeah. I personally think we should rather add a Share button to the whatsnew/firstrun pages to encourage people to tweet "Hey I just installed Firefox!" or something else, but these `og` tags might not be necessary anyway, because Share buttons can have an arbitrary string and URL.

Comment (alexgibson) [inline bedrock/firefox/templates/firefox/whatsnew-base-resp.html]: Ah, @kyoshino I see you have changed the sharing URL for these pages? That makes me not so worried :+1: 

Still, I think Flod's comments about this being an edge case still warrant some investigation.

Comment (alexgibson) [inline bedrock/firefox/templates/firefox/whatsnew-base-resp.html]: Well, seeing as there's probably very little traffic to these pages anyway, changing them now still seems maybe unnecessary.

Comment (alexgibson) [inline bedrock/firefox/templates/firefox/whatsnew-base-resp.html]: @kyoshino why not just add these meta tags to the most recent /whatsnew and /firstrun? Seems that would simplify both the string situation and not make the templates less confusing?

Comment (kyoshino) [inline bedrock/firefox/templates/firefox/whatsnew-base-resp.html]: Actually I cannot recognize which are the latest whatsnew/firstrun pages :confused: And if new variations of those pages is added, we could miss the string, so I thought the intermediate `whatsnew-base` template was a good idea.

Comment (alexgibson) [inline bedrock/firefox/templates/firefox/whatsnew-base-resp.html]: Using a base template for these strings might be a fine enough idea, but it just seems like changing all these old templates is a wasted effort?

Whatsnew: https://github.com/mozilla/bedrock/blob/master/bedrock/firefox/views.py#L472

Firstrun: https://github.com/mozilla/bedrock/blob/master/bedrock/firefox/views.py#L418

File: bedrock/firefox/templates/firefox/nightly_firstrun.html [hunk 7]
```diff
@@ -4,9 +4,7 @@
 
 {# firstrun for nightlies #}
 
-{% extends "/firefox/base-resp.html" %}
-
-{% block extra_meta %}<meta name="robots" content="noindex">{% endblock %}
+{% extends "/firefox/whatsnew-base-resp.html" %}
```
Comment (alexgibson) [inline bedrock/firefox/templates/firefox/nightly_firstrun.html]: Also seems confusing that firstrun templates are now extending `whatsnew-base-resp.html`. Should probably find a better name?

File: bedrock/firefox/templates/firefox/nightly_firstrun.html [hunk 8]
```diff
@@ -17,6 +15,8 @@
 {% block page_title_prefix %}{% endblock %}
 {% block page_title %}{{ _('Firefox Nightly First Run Page') }}{% endblock %}
 {% block page_title_suffix %}{% endblock %}
+{% block page_og_url %}http://nightly.mozilla.org/{% endblock %}
```
Comment (alexgibson) [inline bedrock/firefox/templates/firefox/nightly_firstrun.html]: `https://`

**Comment by kyoshino** [comment]
Sorry for the delay here. I have removed the intermediate whatsnew templates and added the `page_og` blocks only to the recent whatsnew/firstrun pages. Does it work? @flodolo

File: bedrock/firefox/templates/firefox/privacy_tour/index.html [hunk 9]
```diff
@@ -11,6 +11,23 @@
 
 {% block extra_meta %}<meta name="robots" content="noindex">{% endblock %}
 
+{#- This will appear as <meta property="og:title"> which can be used for social share -#}
```
Comment (alexgibson) [inline bedrock/firefox/templates/firefox/privacy_tour/index.html]: This is an old page, so this one can probably just be removed.

File: bedrock/firefox/templates/firefox/search_tour/index.html [hunk 10]
```diff
@@ -6,6 +6,23 @@
 
 {% block extra_meta %}<meta name="robots" content="noindex">{% endblock %}
 
+{#- This will appear as <meta property="og:title"> which can be used for social share -#}
```
Comment (alexgibson) [inline bedrock/firefox/templates/firefox/search_tour/index.html]: Same here

File: bedrock/firefox/templates/firefox/privacy_tour/index.html (hunk 9 above)
Comment (kyoshino) [inline bedrock/firefox/templates/firefox/privacy_tour/index.html]: Will remove it.

File: bedrock/firefox/templates/firefox/dev-firstrun.html (hunk 3 above)
Comment (alexgibson) [inline bedrock/firefox/templates/firefox/dev-firstrun.html]: Since this is a `/firstrun` page we should probably override `{% block page_og_url %}` too.

Comment (alexgibson) [inline bedrock/firefox/templates/firefox/dev-firstrun.html]: Oh, nevermind - that's already done here below :+1:

File: bedrock/mozorg/helpers/misc.py [hunk 11]
```diff
@@ -462,8 +462,10 @@ def absolute_url(url):
 
     if url.startswith('//'):
         prefix = 'https:'
-    else:
+    elif url.startswith('/'):
```
Comment (alexgibson) [inline bedrock/mozorg/helpers/misc.py]: This is only a tiny change and has a test below, so looks OK to me. But I'd appreciate it if either @pmclanahan or @jgmize could take a quick look and give their blessing.

File: bedrock/firefox/templates/firefox/australis/fx36/tour-base.html [hunk 12]
```diff
@@ -8,6 +8,23 @@
 
 {% block extra_meta %}<meta name="robots" content="noindex">{% endblock %}
 
+{#- This will appear as <meta property="og:title"> which can be used for social share -#}
+{% block page_og_title %}
+  {% if l10n_has_tag('meta_update_201503') %}
```
Comment (flodolo) [inline bedrock/firefox/templates/firefox/australis/fx36/tour-base.html]: Since we're using these 2 strings in 3 files, it makes more sense to have them in a shared file (and we can't have tags on those).

So, let's remove the l10n_has_tag conditional on these pages
```
bedrock/firefox/templates/firefox/australis/fx36/tour-base.html
bedrock/firefox/templates/firefox/australis/tour-base.html
bedrock/firefox/templates/firefox/new.html
```
I'll move the strings currently in firefox/new.lang into main.lang

File: bedrock/mozorg/helpers/misc.py (hunk 11 above)
Comment (pmclanahan) [inline bedrock/mozorg/helpers/misc.py]: I actually don't see a test for the new "else" branch of this code. Should have a unit test in there to ensure what you expect happens when the function is passed something that doesn't start with a slash.

Comment (kyoshino) [inline bedrock/mozorg/helpers/misc.py]: Umm, the added `test_full` didn't make sense. I have added `test_urls` instead to check URLs as intended.

File: bedrock/firefox/templates/firefox/australis/fx36/tour-base.html (hunk 12 above)
Comment (kyoshino) [inline bedrock/firefox/templates/firefox/australis/fx36/tour-base.html]: Thanks @flodolo, `main.lang` might be the best place for those strings.

Comment (alexgibson) [inline bedrock/firefox/templates/firefox/australis/fx36/tour-base.html]: @kyoshino - I think we're good to merge this one once we remove the l10n_has_tag conditional in the pages @flodolo suggests. Thanks for all your work on this one!

File: bedrock/mozorg/helpers/misc.py (hunk 11 above)
Comment (alexgibson) [inline bedrock/mozorg/helpers/misc.py]: @pmclanahan appreciative if you could take a quick look at the updated tests here :+1:

File: bedrock/firefox/templates/firefox/australis/fx36/tour-base.html (hunk 12 above)
Comment (kyoshino) [inline bedrock/firefox/templates/firefox/australis/fx36/tour-base.html]: Removed the `l10n_has_tag` conditional.

File: bedrock/mozorg/helpers/misc.py (hunk 11 above)
Comment (pmclanahan) [inline bedrock/mozorg/helpers/misc.py]: APPROVED :+1:
